frontend/test/frontend_tests/logic/copying_and_duplicating_test.cljs

```diff
@@ -0,0 +1,400 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;;
+;; Copyright (c) KALEIDOS INC
+
+(ns frontend-tests.logic.copying-and-duplicating-test
+  (:require
+   [app.common.test-helpers.components :as cthc]
+   [app.common.test-helpers.compositions :as ctho]
+   [app.common.test-helpers.files :as cthf]
+   [app.common.test-helpers.ids-map :as cthi]
+   [app.common.test-helpers.shapes :as cths]
+   [app.common.uuid :as uuid]
+   [app.main.data.workspace :as dw]
+   [app.main.data.workspace.colors :as dc]
+   [app.main.data.workspace.libraries :as dwl]
+   [app.main.data.workspace.selection :as dws]
+   [cljs.test :as t :include-macros true]
+   [frontend-tests.helpers.pages :as thp]
+   [frontend-tests.helpers.state :as ths]))
+
+(t/use-fixtures :each
+  {:before thp/reset-idmap!})
+
+;; Related .penpot file: common/test/cases/copying-and-duplicating.penpot
+(defn- setup-file []
+  (-> (cthf/sample-file :file1 :page-label :page-1)
+      (ctho/add-simple-component :simple-1 :frame-simple-1 :rect-simple-1 :child-params {:type :rect :fills (cths/sample-fills-color :fill-color "#2152e5") :name "rect-simple-1"})
+
+      (ctho/add-frame :frame-composed-1 :name "frame-composed-1")
+      (cthc/instantiate-component :simple-1 :copy-simple-1 :parent-label :frame-composed-1 :children-labels [:composed-1-simple-1])
+      (cths/add-sample-shape :rect-composed-1 :parent-label :frame-composed-1 :fills (cths/sample-fills-color :fill-color "#B1B2B5"))
+      (cthc/make-component :composed-1 :frame-composed-1)
+
+      (ctho/add-frame :frame-composed-2 :name "frame-composed-2")
+      (cthc/instantiate-component :composed-1 :copy-composed-1-composed-2 :parent-label :frame-composed-2 :children-labels [:composed-1-composed-2])
+      (cthc/make-component :composed-2 :frame-composed-2)
+
+      (cthc/instantiate-component :composed-2 :copy-composed-2)
+
+      (ctho/add-frame :frame-composed-3 :name "frame-composed-3")
+      (ctho/add-group :group-3 :parent-label :frame-composed-3)
+      (cthc/instantiate-component :composed-2 :copy-composed-1-composed-3 :parent-label :group-3 :children-labels [:composed-1-composed-2])
+      (cths/add-sample-shape :circle-composed-3 :parent-label :group-3 :fills (cths/sample-fills-color :fill-color "#B1B2B5"))
+      (cthc/make-component :composed-3 :frame-composed-3)
+
+      (cthc/instantiate-component :composed-3 :copy-composed-3 :children-labels [:composed-2-composed-3])
+      (cthf/add-sample-page :page-2)
+      (cthf/switch-to-page :page-1)))
+
+
+(defn- copy-paste-shape
+  ([id file]
+   (copy-paste-shape id file nil nil))
+  ([id file target-page-label target-container-id]
```

Comment:
I'd prefer to use optional parameters with `&` than to create different arities. I think this other way is clearer and more idiomatic, and also more flexible (you can choose different sets of optional arguments when calling).